README
#$Id: README 1506 2013-09-06 01:10:34Z jimk $
List::Compare - Compare elements of two or more lists

This document refers to version 0.38 of List::Compare.  This version was
released September 05, 2013.

To install this module on your system, place the tarball archive file in a 
temporary directory and call the following:

% gunzip List-Compare-0.38.tar.gz
% tar xf List-Compare-0.38.tar
% cd List-Compare-0.38
% perl Makefile.PL
% make
% make test
% make install

If during installation you wish to view more information on test results, 
substitute the fllowing for the sixth line in the sequence of commands
above:

% make test TEST=VERBOSE

If you are installing this module over any earlier version, you may 
substitute the following for the last line in the sequence of commands 
above:

% make install UNINST=1

If you are installing this module on a Win32 system with 'nmake', 
substitute 'nmake' for 'make' in the sequence of commands above.

The author has found that trying to install this distribution with 
(a) older (pre-6.16) versions of ExtUtils::MakeMaker, (b) on older 
versions of Perl (e.g., 5.6.0), and (c) on older Linux distributions 
(e.g., RedHat Linux 7.3) can result in a situation where the module's 
Plain Old Documentation, when transformed into manual pages, is not 
automatically placed in the optimal location for reading thru 'man' 
and 'info' (even though it reads perfectly through 'perldoc').  If you 
experience this problem, issue the following at the command line 
(adapted from a suggestion by Michael Schwern on the module-authors 
mailing list):

% perl Makefile.PL INSTALLSITEMAN3DIR=/usr/share/man/man3

List::Compare uses the Carp module which is part of the standard Perl 
distribution.  Other than that, there are no module dependencies in 
this version of List::Compare.

Certain methods in this version of List::Compare are included solely 
for backwards compatibility with earlier versions and are deprecated.  
When called, they print warning messages via Carp.  In earlier versions 
of List::Compare these warning messages would appear when you called 
'make test' as part of the installation process.  This was harmless 
but annoying and has been fixed.

Beginning with version 0.25 in April 2004, there is a modification to the 
interface of two methods/functions: are_members_which() and 
are_members_any().  Whereas previously the strings to be tested could be 
passed to the subroutine either as a flat list or via a reference to an 
anonymous array, now those items must be passed via reference to an 
anonymous array.

Beginning with version 0.26 in April 2004, the functionality previously 
found in List::Compare::SeenHash has been incorporated directly into 
List::Compare.  Hence, List::Compare::SeenHash and the test suite files 
associated with it have been deprecated and are no longer included in the 
CPAN distribution.  Please see the List::Compare documentation for 
further details.

In sending e-mail to the author, please put "List::Compare" or 
"List-Compare" in the subject line.

Author:  James E. Keenan (jkeenan@cpan.org).
Originally created May 20, 2002.

Copyright (c) 2002-08 James E. Keenan.  United States.  All rights reserved. 
This is free software and may be distributed under the same terms as Perl
itself.



Changes
#$Id: Changes 1506 2013-09-06 01:10:34Z jimk $
Revision history for Perl extension List::Compare.

0.04  Sat Jun  8 2002
	- last draft version before using h2xs

0.05  Sun Jun  9 2002
	- original version; created by h2xs 1.20 with options
		-XA -n List::Compare

0.06  Wed Jun 12 2002
	- documentation changes; POD broke

0.07  Thu Jun 13 2002
	- documentation changes

0.08  Sat Jun 15 2002
	- added method get_version()
	- used non-idiomatic syntax for 'foreach' in hope of making module more backwardly compatible
	- to free up memory, zeroed out %seenA and %seenB when no longer needed
	- added method get_bag()
	- inspired by Set::Scalar, used typeglob syntax to implement aliases for methods
	- revised POD extensively

0.09  Sun Jun 16 2002
	- added tests to test get_bag()
	- implemented Carp module for warnings

0.10  Wed Jun 19 2002
	- corrected one typo in documentation
	- commented out # require 5.00562 and use warnings
	- first version uploaded to CPAN

0.11  Thu Jun 20 2002
	- corrected errors in POD that caused HTML to display incorrectly

0.12  Sat Jul 13 2002
	- major expansion
	- capability to accelerate processing of comparisons of 2 lists 
	  when only 1 comparison method is expected to be called
	- capability to compare more than 2 lists at a time
	- introduction of methods including:
		get_shared()
		get_nonintersection()
		print_subset_map()
		print_equivalence_map()
	- not released to CPAN due to errors in README and Changes

0.13  Sat Aug 03 2002
	- corrected documentation errors found in v0.12
	- released to CPAN

0.14  Sun Aug 18 2002
	- commented out 'use warnings;' to permit module to run on pre-5.6 versions
	- corrected one error in POD
	- removed e-mail addresses of authors of other modules from POD

0.15  Sat Sep 07 2002
	- conformed CPAN addresses of other modules to new structure of search.cpan.org

0.16  Sat Mar 08 2003
	- for each method which returned an array, added an analogous method which
	  returns a reference to that array

0.17  Thu May 22 2003
	- eliminated certain synthetic variables
	- created List::Compare::SeenHash

0.18  Sun Jun 1 2003
	- added methods member_which(), single_member_which()

0.19  Sun Jun 1 2003
	- added methods member_any(), single_member_any()
	- fixed annoying carping messages when running 'make test'

0.20 Fri Jun 6 2003
	- updated documentation discussing changes since v0.17
	- changed names of recently added methods to is_member_which(), is_member_which_ref(),
	  are_members(), is_member_any(), are_members_any()

0.21 Sun Oct 26 2003
	- added functional interface to list comparison functions:  List::Compare::Functional; this required changes in List::Compare::Base::Accelerated and the creation of List::Compare::Base::_Engine

0.22 Sun Nov 23 2003
	- corrected conceptual flaw in List::Compare::SeenHash so as to exclude values in a seen-hash mathematically equal to zero
	- implemented 'unsorted' option for list comparisons

0.23 [Accidentally skipped.]

0.24 Sun Mar 28 2004
	- first implementation of 'multiple accelerated' mode wherein user can request speedier processing of more than two lists when only one comparison among the lists is sought
	- converted test suite to use Test::Simple

0.25 Sun Apr 4 2004
	- continued implementation of 'multiple accelerated' mode; applied to List::Compare::SeenHash and List::Compare::Functional
	- modification of interfaces to methods/functions are_members_which() and are_members_any throughout; items to be tested must be passed (by reference) as an anonymous array rather than as a flat list

0.26  Sun Apr 11 2004
        - List::Compare::SeenHash and associated tests deprecated
        - capacity to pass references to seen-hashes to constructor now 
transferred directly into List::Compare
	- corrections to POD

0.27  Sun Apr 18 2004
	- corrections to POD

0.28  Sun Apr 25 2004
	- implementation of is_LdisjointR
	- correction of error in tests of multiple-accelerated mode
	- correction of error in internal subroutine _argument_checker_0
	- further implication of deprecation of List::Compare::SeenHash is that all the code in List::Compare::Base::Regular, L:C:B:Accelerated, L:C:B:Multiple and L:C:B:Multiple::Accelerated is no longer shared between List::Compare and List::Compare::SeenHash and hence can be taken back into List::Compare; thereby eliminating the 'use base' calls from List::Compare.

0.29  Sun May 16 2004
	- implementation of alternative ('Adlerian') interface wherein a single hash reference is passed to constructor or function 

0.30  Fri May 21 2004
	- documentation corrections only

0.31  Sun Aug 15 2004
	- added methods get_unique_all() and get_complement_all(); added 3335 
tests; repositioned most methods in List::Compare::Base::_Engine to 
List::Compare::Accelerated as they were no longer also being used in 
List::Compare::Functional; other internal code clean-up

0.32  Sun Sep 18 2005
    - No changes in functionality.  Distribution restructured so that modules
      are now underneath the 'lib/' directory.

0.33  Sun Feb 19 2006
    - Documentation changes only.  Support for List-Compare is now offered via
      dedicated mailing list.

0.34 Mon Nov  5 18:35:15 EST 2007
    - Add require 5.6 to Makefile.PL.

0.35 Sun May 18 19:54:32 EDT 2008
    - Fix bug reported by Charles Bailey.  In object-oriented interface, two
      lists case, get_shared and get_shared_ref should default to
      get_intersection and get_intersection_ref.
    - Complete overhaul of test suite.  Tests are now grouped by interface in
      a more logical order.

0.36 Fri May 23 21:41:55 EDT 2008
    - IO::CaptureOutput was mispositioned:  should have been directly under t/
      rather than under t/Test/.

0.37 Sat Jun  7 19:46:52 EDT 2008
    - Tests refined and code refactored so as to achieve 100% subroutine,
      statement, branch and condition coverage by test suite as measured by
      Devel::Cover.  Some cleanup of POD.

0.38 Thu Sep  5 21:09:10 EDT 2013

lib/List/Compare/Base/_Engine.pm
package List::Compare::Base::_Engine;
#$Id: _Engine.pm 1506 2013-09-06 01:10:34Z jimk $
$VERSION = 0.38;
# Holds subroutines used within 
# List::Compare::Base::Accelerated and List::Compare::Functional
# As of:  09/18/2005
use Carp;
use List::Compare::Base::_Auxiliary qw(
    _equiv_engine 
    _calculate_seen_xintersection_only
    _calculate_union_seen_only
);
@ISA = qw(Exporter);
@EXPORT_OK = qw|
    _unique_all_engine
    _complement_all_engine
|;
use strict;
local $^W = 1;

sub _unique_all_engine {
    my $aref = shift;
    my ($seenref, $xintersectionref) = 
        _calculate_seen_xintersection_only($aref);
    my %seen = %{$seenref};
    my %xintersection = %{$xintersectionref};

    # Calculate @xunique
    # Inputs:  $aref    %seen    %xintersection
    my (@xunique);
    for (my $i = 0; $i <= $#{$aref}; $i++) {
        my %seenthis = %{$seen{$i}};
        my (@uniquethis, %deductions, %alldeductions);
        # Get those elements of %xintersection which we'll need 
        # to subtract from %seenthis
        foreach (keys %xintersection) {
            my ($left, $right) = split /_/, $_;
            if ($left == $i || $right == $i) {
                $deductions{$_} = $xintersection{$_};
            }
        }
        foreach my $ded (keys %deductions) {
            foreach (keys %{$deductions{$ded}}) {
                $alldeductions{$_}++;
            }
        }
        foreach (keys %seenthis) {
            push(@uniquethis, $_) unless ($alldeductions{$_});
        }
        $xunique[$i] = \@uniquethis;
    }
    return \@xunique;
}

sub _complement_all_engine {
    my ($aref, $unsortflag) = @_;
    my ($unionref, $seenref) = _calculate_union_seen_only($aref);
    my %seen = %{$seenref};
    my @union = $unsortflag ? keys %{$unionref} : sort(keys %{$unionref});

    # Calculate @xcomplement
    # Inputs:  $aref @union %seen
    my (@xcomplement);
    for (my $i = 0; $i <= $#{$aref}; $i++) {
        my %seenthis = %{$seen{$i}};
        my @complementthis = ();
        foreach (@union) {
            push(@complementthis, $_) unless (exists $seenthis{$_});
        }
        $xcomplement[$i] = \@complementthis;
    }
    return \@xcomplement;
}

1;


__END__

=head1 NAME

List::Compare::Base::_Engine - Internal use only

=head1 VERSION

This document refers to version 0.38 of List::Compare::Base::_Engine.
This version was released September 05, 2013.

=head1 SYNOPSIS

This module contains subroutines used within List::Compare and 
List::Compare::Functional.  They are not intended to be publicly callable.

=head1 AUTHOR

James E. Keenan (jkeenan@cpan.org).  When sending correspondence, please 
include 'List::Compare' or 'List-Compare' in your subject line.

Creation date:  May 20, 2002.  Last modification date:  September 05, 2013. 
Copyright (c) 2002-04 James E. Keenan.  United States.  All rights reserved. 
This is free software and may be distributed under the same terms as Perl
itself.

=cut 

